Reland: [TensorExpr] Add methods for inspecting generated code in `TensorExprKernel`.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57560 Reland: [TensorExpr] Add methods for inspecting generated code in `TensorExprKernel`.**
* #57553 Reland: [StaticRuntime] Use NNC's call_raw API to reduce call overheads.
* #57552 Reland: [TensorExpr] Add `TensorExprKernel::runFast` method.
* #57551 Reland: [TensorExpr] Add `CodeGen::call_raw` method.

The new methods allow to peak into bufferArgs which describe parameters
that codegen expects. This description includes info whether a given
parameter is a scalar var or a buffer and in case it's a buffer allows
to get the corresponding `Buf*` pointer from which we could get the
expected sizes.

Relanding #57074  which was reverted because I forgot to guard a new
test with `ifdef LLVM`.

Differential Revision: [D28199048](https://our.internmc.facebook.com/intern/diff/D28199048)

Commit:
Reland: [TensorExpr] Add methods for inspecting generated code in `TensorExprKernel`.

The new methods allow to peak into bufferArgs which describe parameters
that codegen expects. This description includes info whether a given
parameter is a scalar var or a buffer and in case it's a buffer allows
to get the corresponding `Buf*` pointer from which we could get the
expected sizes.

Relanding #57074  which was reverted because I forgot to guard a new
test with `ifdef LLVM`.

[ghstack-poisoned]

navahgar commented on test/cpp/tensorexpr/test_kernel.cpp:
```diff
@@ -1288,5 +1288,49 @@ TEST_F(Kernel, RunFast) {
+  // Check that we could retrieve info about codegen parameters
+  auto constants = k.getConstantDescriptors();
+  auto buf_args = k.getBufferArgs();
+  // Expected buf args: [input0, output0, constant0]
```
Is it possible to look for these specific buf args here?

ZolotukhinM replied:
From the perspective of this API the exact Buf pointers are not really useful - what's important is that we can access them (the check whether they are buffers and not scalar variables is mostly a smoke check). We have plenty of other tests that check that the pointers themselves are correct - every test running a kernel exercises that part.

navahgar reviewed (approved):
LGTM. One minor comment below.

Commit:
rebase on "Reland: [TensorExpr] Add methods for inspecting generated code in `TensorExprKernel`."


The new methods allow to peak into bufferArgs which describe parameters
that codegen expects. This description includes info whether a given
parameter is a scalar var or a buffer and in case it's a buffer allows
to get the corresponding `Buf*` pointer from which we could get the
expected sizes.

Relanding #57074  which was reverted because I forgot to guard a new
test with `ifdef LLVM`.

Differential Revision: [D28199048](https://our.internmc.facebook.com/intern/diff/D28199048)

[ghstack-poisoned]
